### PR TITLE
WIP multi runbook support

### PR DESCRIPTION
**Pull request by doramatadora**
*Work In Progress: feel free to ignore notifications*
Description / docs to follow

- multiple runbook support in repos 
- repository application config (via `.github/runbooks.yml` which must exist on the master branch)
- automated biz-ops updates for apps with manual release processes

Example output: https://github.com/Financial-Times/runbook.md/runs/199138210

### CODE REVIEW

File: lambdas/ingester/src/commands/webhook/runbookCheck.js [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+					runbookRx.test(path) &&
+					!this.config.isPathExcluded(path),
+			);
+		}
+
+		return this.runbooks;
+	}
+
+	async processRunbooks(context) {
+		if (!this.runbooks) {
```
Comment (wheresrhys) [inline lambdas/ingester/src/commands/webhook/runbookCheck.js]: is this and the one on line 62 needed?

File: lambdas/ingester/src/commands/webhook/runbookCheck.js [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+				systemCode: runbook.systemCode,
+			},
+		});
+
+		const result = {
+			state: 'success',
+			...runbook,
+		};
+
+		const writeToBizOps = this.config.writeToBizOpsEnabled(
```
Comment (wheresrhys) [inline lambdas/ingester/src/commands/webhook/runbookCheck.js]: shouldWriteToBizOps?

writeToBizOps reads like the name of a method that actively writes to Biz Ops

File: lambdas/ingester/src/commands/webhook/runbookCheck.js [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+			];
+			return false;
+		}
+	}
+
+	// TODO: batch writes for performance
+	// https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/dynamodb-example-table-read-write-batch.html
+	async storeIngestResults() {
+		if (!isArrayNotEmpty(this.runbooks)) {
+			return this.throwCriticalError('No runbooks ingested');
```
Comment (wheresrhys) [inline lambdas/ingester/src/commands/webhook/runbookCheck.js]: Is this sometimes gonna happen in situations where everything has worked ok, but there just are no runbook changes in this PR?

If os, then would a better name for `throwCriticalError` be `bail`. Say I was debugging something in the app in future I would probably read throwCriticalError as a code path that is only supposed to be hit if something's gone drastically wrong

Comment (wheresrhys) [reply lambdas/ingester/src/commands/webhook/runbookCheck.js]: Also, the `No runbooks ingested` could say whether tha's an expected & ok state of affairs or not

File: lambdas/ingester/src/statusPageHandler.js [hunk 4]
```diff
@@ -1,65 +1,68 @@
...
+		reporter: 'STATUS_PAGE',
+		repository,
+		runbookSha: hash,
+		commitSha,
+	});
+
 	try {
-		const storedResult = await getStoredResult(repository, hash);
+		const [storedResult, { checkRunUrl }] = await Promise.all([
+			getStoredResult(repository, hash),
```
[the remark this answers is not in the document]
Comment (doramatadora) [reply lambdas/ingester/src/statusPageHandler.js]: The hash is now the runbook file hash, which will work for all cases.

[the remark this answers is not in the document]
Comment (doramatadora) [reply lambdas/ingester/src/statusPageHandler.js]: The hash is now the runbook hash. System codes only exist in the (rare, we hope) contexts of multi-runbook repos & writing to biz-ops for manual release processes, so we don't want to rely on them.

File: lambdas/ingester/src/releaseLogHandler.js [hunk 5]
```diff
@@ -204,12 +152,15 @@ const ingestRunbookMDs = (runbookMDs, childLogger) =>
...
-						bizOpsApiKey,
-					});
+					const result = await ingest(
+						{
+							systemCode,
+							content,
+							writeToBizOps: true,
+							bizOpsApiKey,
+						},
+						userName,
```
[the remark this answers is not in the document]
Comment (doramatadora) [reply lambdas/ingester/src/releaseLogHandler.js]: I'm going to root out all references to username, actually – it's only used for a header that we are deprecating, as per your comment about the manual form.

File: lambdas/ingester/src/lib/external-apis.js [hunk 6]
```diff
@@ -24,14 +25,15 @@ const callExternalApi = async ({
 		...options,
 		agent: keepAliveAgent,
 	});
-	if (!expectedStatuses.includes(fetchResponse.status)) {
+	const { status, statusText } = fetchResponse;
+	if (!expectedStatuses.includes(status)) {
 		logger.error(
 			{ event: `Attempt to ${method} to ${url}`, options },
```
[the remark this answers is not in the document]
Comment (doramatadora) [reply lambdas/ingester/src/lib/external-apis.js]: The hazards of refactoring bits of code and not the entire file...

File: lambdas/ingester/src/commands/ingest.js [hunk 7]
```diff
@@ -26,29 +26,33 @@ const decorateError = props => {
...
+const ingest = async (payload, username) => {
 	const {
 		content: rawRunbook,
 		writeToBizOps,
 		systemCode,
 		bizOpsApiKey,
 	} = payload;
 	if (!rawRunbook) {
-		throw decorateError({ message: 'Please supply RUNBOOK.md content' });
+		throw decorateError({
```
[the remark this answers is not in the document]
Comment (doramatadora) [reply lambdas/ingester/src/commands/ingest.js]: I disagree with `event` – this is just an API so I'd rather leave it up to the client (`ingest` caller) to log an event for the execution context based on a code. Not all errors get logged and the error.code is tied to different events in different contexts (e.g. release log vs webhook handling)

File: lambdas/ingester/src/lib/code-validation.js [hunk 8]
```diff
@@ -120,9 +120,9 @@ const transformCodesIntoNestedData = async (username, data) => {
 	}
 	const { query, propertyMappings } = buildGraphQLQuery(uniqueBizOpsCodes);
 	const { json: bizOpsResponse } = await queryBizOps(
-		username,
 		process.env.BIZ_OPS_API_KEY,
```
[the remark this answers is not in the document]
Comment (doramatadora) [reply lambdas/ingester/src/lib/code-validation.js]: Tidied :) No reason to keep it

**Pull request merged**